marshmallow/schema.py

```diff
@@ -706,10 +701,32 @@ def _update_fields(self, obj=None, many=False):
         else:
             field_names = self.set_class(self.declared_fields.keys())
 
+        declared_fields = field_names
+        invalid_fields = self.set_class()
+        if self.only is not None:
+            # Return only fields specified in only option
+            only = self.set_class(self.only)
+            field_names = only
+            invalid_only = only - declared_fields
+            invalid_fields |= invalid_only
+
         # If "exclude" option or param is specified, remove those fields
         excludes = set(self.opts.exclude) | set(self.exclude)
         if excludes:
             field_names = field_names - excludes
+            invalid_excludes = excludes - declared_fields
+            invalid_fields |= invalid_excludes
+
+        if invalid_fields:
+            invalid_names = ['"{}"'.format(name) for name in invalid_fields]
+            invalid_names = utils.iterable_to_string(invalid_names)
+            if len(invalid_fields) > 1:
+                message = '{0} are not valid fields for {1}.'
+            else:
+                message = '{0} is not a valid field for {1}.'
+            message = message.format(invalid_names, self.__class__.__name__)
+            raise AttributeError(message)
```

Comment:
`AttributeError` doesn't feel like the correct error. From the [docs](https://docs.python.org/3/library/exceptions.html#AttributeError).

> Raised when an attribute reference (see Attribute references) or assignment fails.

I think `ValueError` is more appropriate.

Reply:
That makes sense. I was thinking of fields as schema class attributes, but that isn't really how they are used.